### ISSUE TITLE
Translation management system

### ISSUE THREAD

**Issue by ctrlcctrlv** [fontforge/fontforge#3814]
FontForge has a bunch of old translations. As seen in #3797/#3808 (etc.), non-experts have a hard time learning how to donate translations to the project, and it takes up developer time helping them.

There are several better ways:

* **Transifex** &mdash; I have experience with this one from an old job of mine. They used to allow just anyone to make an account if you promised you were an open source project, but it seems that they are getting less and less generous. @fitojb, one of our translators, no longer works with them. Their website is snazzy, but FontForge needs something that's going to last, and which our translators like, so pass.
* **Translatewiki** &mdash; This is my preferred option. It is used by several high priority FOSS projects and websites, including OpenStreetMap, Internet Archive, MediaWiki and MathJax., I talked to the owner of it in IRC and we don't have to do much to get started. We'd need to:
  * change how `FontForge.pot` is built, building it as part of the normal build process. Us developers need to just be good about checking in new strings when we add them.
  * give a translatewiki team member push access to this repository so they can update the `po` files.
  * write a quick article about FontForge and explain some of the terminology (I'll do this part)
* **Weblate** &mdash; recommended by @fitojb, I don't know anything about it. But let's put it on the list anyway. :-)

**Comment by frank-trampe** [comment]
I'd usually prefer Python over PHP, but it looks like we get much better human support for Translatewiki, so I'd be inclined to go that route.

Feedback from @jtanx, @davelab6, and @JoesCat would be much appreciated.

**Comment by jtanx** [comment]
> give a translatewiki team member push access to this repository

So changes are just committed directly to master? Initial impression is I'm not sure how I feel about that.

**Comment by Nikerabbit** [comment]
> So changes are just committed directly to master? Initial impression is I'm not sure how I feel about that.

Hi, I am from translatewiki.net. Can you elaborate what makes you feel uneasy?

**Comment by jtanx** [comment]
Yep, usually just because all changes to master goes through pull requests, and it feels like a little bit of a security hole to allow otherwise.

**Comment by nemobis** [comment]
Jeremy Tan, 22/07/19 12:13:
> Yep, usually just because all changes to master goes through pull 
> requests, and it feels like a little bit of a security hole to allow 
> otherwise.

Translations as data are not typically executable.

While some malformed translations may be submitted by translators, using 
a web-based translator greatly reduces the chances of errors compared to 
the manual systems your translators seem to be using now. Translate also 
allows to code checkers which either warn or block users upon errors.

**Comment by ctrlcctrlv** [comment]
@frank-trampe Having worked with PHP I also don't like it, but using Translatewiki doesn't require us to know anything about PHP.

@jtanx I understand your concern, but I tend to trust them. They manage translations for a bunch of large projects and are only pushing strings, not code. It's true one of them could go rogue - but so could any of us, this project has an unusually large number of (both active and inactive) contributors with push access.

**Comment by jtanx** [comment]
It's not so much the translations that I'm thinking about; giving commit access allows you to change *any* part of the repo.

Pushing directly to master makes this way harder to audit; sure, I'm almost certain you won't do anything nefarious, but now I'm having to keep on top of what was committed if I ever wanted to make sure that was the case.

On the other hand, the onus is on us to merge pull requests, which makes that much easier.

Having said all that, I'm not sure how Transifex works either, so maybe it has similar requirements.

Have you ever thought about setting up a bot that automates creating pull requests?

**Comment by jtanx** [comment]
> It's true one of them could go rogue - but so could any of us, this project has an unusually large number of (both active and inactive) contributors with push access.

And yeah, definitely concede that.

**Comment by ctrlcctrlv** [comment]
@Nikerabbit Can your team open pull requests instead of directly pushing to master?

If not, what about if we put the `po` directory into a submodule under the fontforge organization, then changed our Git build instructions to accommodate that? We'd let you push to `fontforge/po`, not `fontforge/fontforge`, and when we build releases, we'd check out the `po` submodule.

**Comment by ctrlcctrlv** [comment]
@jtanx If they can do the fontforge/po thing, would you be ok with that? I can open a PR to change what autotools is doing and fix up the developer documentation (CONTRIBUTING.md, INSTALL-git.md) to account for that.

https://git-scm.com/book/en/v2/Git-Tools-Submodules

**Comment by Nikerabbit** [comment]
We do not currently support GitHub pull requests. We have a great desire towards full automation, and someone having to manually approve the pull requests goes against that. What would you review on those pull requests? Translation quality can only be reviewed if you speak the languages. That's better done by translators in translatewiki.net. The po files will be generated by our code, so par any bugs, they are guaranteed to be valid. In addition to that, we have automated checks that can warn about translation mistakes, or even prevent saving those in the first place. At the time of project setup, we will also check what automated checks should be enabled.

Changes to our code go through a peer review using Gerrit. Our ssh key is protected with passphrase (known to roots) and it is used through keyholder proxy that does not expose to the key to the user during exports.

Basically, it requires you to trust our roots, our translation exporters, our hosting provider, our code review process and our ability to keep malicious people out of our servers.

We currently support the following push models (let's say A is the branch where development happens, usually `master` and B is another branch):
* Read pot from A, push translations to A
* Read pot from A, force push translations to B (rebased on A)
* Read pot from A, push translations to B (not rebased)

We do not support cross-repo (read pot from repo X, push translations to repo Y).

**Comment by skef** [comment]
That seems fine then. Github has branch-based permissions. Just give the tool access to a branch and periodically pull the results into `master` when it makes sense.

**Comment by ctrlcctrlv** [comment]
First off, thank you so much @Nikerabbit for your patience. You've had the patience of a saint chatting with me in IRC while I bounce ideas off you, and your patience continues and is appreciated here. I have no problem with using branches instead of submodules :)

**Comment by Nikerabbit** [comment]
It does have the downside that it requires manual work from someone (and it's even less obvious than pull requests), but if you can commit to doing that regularly, I am okay with that.

We care strongly that the translations do reach the end users. We do drop projects where this doesn't happen, but usually it is the case that the project has gone inactive and stopped doing releases.

**Comment by skef** [comment]
I don't think there will be a problem committing to pulling as frequently as there are new "official" releases.

**Comment by ctrlcctrlv** [comment]
We do about two releases per year, I'd personally like us to do four (one per quarter), but no matter what the schedule, there's already much that needs to be done about release time; we can easily put on the list "merge `translatewiki` branch to `master` branch to update translations".

**Comment by khaledhosny** [comment]
> Translations as data are not typically executable.

Bad gettext translation (e.g. bad printf placeholder) can and does crash applications. Linting tools can catch some/most of these but they can be fooled often easily (having done this myself plenty of times when they incorrectly flagged correct translation).

**Comment by JoesCat** [comment]
I'm okay with where this conversation is headed.
@ctrlcctrlv is familiar with the process and holds lead direction,
@frank-trampe and @jtanx hold ultimate yes/no on decision to go with it.
translatewiki certainly would not want to tarnish it's reputation with rogue problems, plus we all want to spread-out the available work - so I think this may work out fine.

@Nikerabbit - we also have 3 files in directory desktop which is used by opensource desktop, would we need to do anything special with this? I would prefer to keep it simple.
...and we also hold html files in directory doc, do you think this has any potential?
...and is this translation an all or nothing type of situation, I know with past translations we have had contributors ask, and contribute, this can be for various reasons, plus also I think it is a feather in one's cap (so to speak), which helps the author find work in their field to say "I did this!" - how do we work with a situation such as this - do we continue, do we say no-more, r do we somehow allow it to work but somehow avoid having people trip over each other trying to update or create the same translation?

Our current gettext situation is a somewhat stripped down version which worked fine in the past with some operating systems that had trouble using gettext....where do we go next?

**Comment by Nikerabbit** [comment]
> we also have 3 files in directory desktop which is used by opensource desktop, would we need to do anything special with this?

These are not localisation file formats, so they cannot be translated out of the box. They need go through a string extractor, that extracts the translatable strings to a supported format. Of course, there also needs to be a reverse process that can create the translated versions from translations.

Gettext seems to have support for extracting strings from those files:
* https://www.gnu.org/software/gettext/manual/html_node/AppData.html#AppData
* https://www.gnu.org/software/gettext/manual/html_node/Desktop-Entry.html#Desktop-Entry

> and we also hold html files in directory doc, do you think this has any potential?

Like above, strings would need to be extracted, but for HTML this is very tricky because it has a lot of mark-up so it is hard to extract strings from it automatically. I would recommend using a MediaWiki with the Translate extension to create translatable documentation for your project.

> ..and is this translation an all or nothing type of situation

We implement two-way sync (except pot which is read-only), so we can handle external updates. But we discourage those heavily as if there are updates on both sides, we will not be spending much effort trying to resolve them and instead just pick one version.

> where do we go next?

From our side, we need the pot file and commit access and then we can start setting up the project if you so want. The setup process is documented in https://translatewiki.net/wiki/Setup_of_a_new_project and we will need some additional info to complete it.

> Bad gettext translation (e.g. bad printf placeholder) can and does crash applications. 

That's a good point. For basic string replacements this is easy: just require the same types as in the original text (with care for ordering support). For other formatting things, it can get more complicated.

**Comment by milotype** [comment]
A question from a translators point of view:
How do translators get credited/mentioned on FontForge's side? Is something like below implemented in translatewiki's system?

It is common, to give infos about oneself in the translations. So like others, I usually add my name & email in the `.po`-file header, something like this (sometimes there is a list of translators):
```
# Croatian User Interface strings for FontForge.
# Copyright (C) 2000-2006 by George Williams
# This file is distributed under the same license as the FontForge package.
# George Williams, <pfaedit@users.sourceforge.net>, 2006.
#
# Translators:
# Milo Ivir <mail@milotype.de>, 2019.
#
```
And then further down in the header, which automatically gets inputed when I save the file:
```
msgid ""
msgstr ""
"Project-Id-Version: fontforge\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2019-04-17 07:50+1000\n"
"PO-Revision-Date: 2019-07-22 10:37+0200\n"
"Last-Translator: Milo Ivir <mail@milotype.de>\n"
"Language-Team: \n"
"Language: hr\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n"
"%10<=4 && (n%100<12 || n%100>14) ? 1 : 2);\n"
"X-Generator: Poedit 2.2.3\n"
```

**Comment by milotype** [comment]
Another question from a translators point of view:
What about downloading files for local translation?
Transifex allows both downloading and uploading translations. This is great, as I love to work offline with a tool of my choice.

**Comment by Nikerabbit** [comment]
Here is an example of file managed by translatewiki.net: https://github.com/dissemin/dissemin/blob/master/locale/de/LC_MESSAGES/django.po

The code preserves existing `# Author:` lines. During setup existing notices need to be converted to this format to retain them.

> What about downloading files for local translation?

We have export&import for offline translation in Gettext format. The file is not the same as in git, so it can only be used through translatewiki.net. Working long times offline easily causes conflicts with other translations, so you need to ask permissions first to use the feature.

**Comment by milotype** [comment]
As long as a decision isn't made on this issue, could you guys make sure to have a current `.pot` file in the po-directory? (My experience as a translator is that having a current pot-file, is always better than using one of the existing `.po` translation files)

**Comment by jtanx** [comment]
I'm actually leaning towards Transifex. By the looks of it, we can just automatically generate and push new strings [via Travis](https://docs.transifex.com/transifex-github-integrations/github-tx-client) without having to commit the pot file, *and* it supports automatically creating PRs to update the translations, so everything then almost becomes set and forget.

**Comment by nemobis** [comment]
Jeremy Tan, 08/08/19 14:03:
> I'm actually leaning towards Transifex.

Why use a proprietary software?

**Comment by jtanx** [comment]
I have no qualms using proprietary software if it works, i.e. GitHub is proprietary, and I use Windows.

From what I saw, it seems like Transifex has better integrations with GitHub, and it supports a workflow that doesn't involve (a) committing generated files, and (b) can auto-create PRs with the requisite changes.

**Comment by nemobis** [comment]
Jeremy Tan, 08/08/19 14:31:
> auto-create PRs with the requisite changes.

This reduces quality of translation compared to direct push, which 
facilitates a tighter collaboration between translators and developers.

**Comment by jtanx** [comment]
> This reduces quality of translation compared to direct push, which facilitates a tighter collaboration between translators and developers.

Sorry, I'm not following the relation between performing the updates via a PR leading to worse quality translations.

**Comment by nemobis** [comment]
Jeremy Tan, 08/08/19 14:36:
> Sorry, I'm not following the relation between performing the updates via 
> a PR leading to worse quality translations.

We have a short primer at
https://translatewiki.net/wiki/Translating:Localisation_for_developers

I can recommend more literature if interested.

I assume you don't want to auto-merge the incoming pull requests, so 
there will be delays and edit conflicts. Projects usually get strained 
by the additional cost of such a process.

**Comment by jtanx** [comment]
> I assume you don't want to auto-merge the incoming pull requests, so there will be delays and edit conflicts.

With the rate of PRs that FontForge receives, I can say with high confidence that won't be an issue.

As a counter argument, there's good reasons for everything to go through PRs. Changes are more easily auditable, plus it allows us to run CI *before* it hits master.

**Comment by milotype** [comment]
Just some info …

**About roles at Transifex:**
https://docs.transifex.com/teams/understanding-user-roles

**From a translators point of view:**
The translation system should have the option for downloading the translated file in the application’s final format (.mo for FontForge). This gives the opportunity to test the translations and do corrections. This is esspecialy true for „accelerators” which are mostly noticed once one runs the application.

**My experience with Transifex:**
I’ve translated Scribus (desktop publishing) into croatian. Transifex allows downloading a .ts file, which doesn’t help much, as the file needs to be „converted” to .qm, in order to use it for the application (for that I use Qt-Linguist). But I think it’s nicer if a translator doesn’t have to bother with converting files.
BTW, I have no idea if the above mentioned is specific only to Scribus’ translation file.

**Comment by ctrlcctrlv** [comment]
It is a problem (if you could call it one) unique to Scribus, yes, [as Scribus uses Qt's i18n](https://github.com/scribusproject/scribus/blob/master/scribus/ui/alignselect.cpp#L112), not gettext.

**Comment by frank-trampe** [comment]
Given that the critical path seems to run through the continuous integration process rather than the translation process, I'm inclined to go with whatever @jtanx prefers here.

**Comment by fitojb** [comment]
I thought we implicitly agreed that Transifex was a no-go! I’m not touching that website anymore and attempt to make sense of its atrocious UI.

> As a counter argument, there's good reasons for everything to go through PRs. Changes are more easily auditable, plus it allows us to run CI before it hits master.

You can do the same with Weblate and Crowdin… they can be configured to open PRs instead of direct pushing.

**Comment by milotype** [comment]
**Please, please,** PLACE A CURRENT .POT FILE in FontForge's po-directory!! 

Even if it does't seem to make sense from a programmer's point of view (if I understood jtanx comment in PR #3863 wright), it DOES help me as a translator to update my local translation files automatically with the newest strings.
I work with Poedit (translation programm). I only need to open my local translation I've been working on and then updated that file with a .pot file. The programm then displays which strings have changed and all "allocations" are updated as well (if there have been changes). I don't know if allocations is the correct word – most strings have some hashed text, like this:
```
#: ../desktop/org.fontforge.FontForge.appdata.xml:42
msgid "FontForge developers"
msgstr ""

```

As mentioned before, all GitHub projects where I'm translating and sending my translations through PR's, do provide a .pot file. Sorry I'm persisting on this issue, but as a non-programmer, pot files are currently the best solution for me.

**Comment by jtanx** [comment]
It's a generated file, it's not that hard to just generate it yourself. I'm pretty sure there's even a way to get poedit to just generate it itself too.

As it currently stands, you can run `make FontForge.pot` from the po folder, assuming you have run configure beforehand.

Otherwise you can just invoke xgettext yourself, like

```
xgettext --from-code=utf-8 -k_ -kN_ -kS_ -kU_ -kNU_ -kP_:1,2 --add-comments=GT: -oFontForge.pot ../fontforge/*.c ../fontforgeexe/*.c ../gdraw/*.c ../gutils/*.c ../inc/*.h ../desktop/org.fontforge.FontForge.appdata.xml
```

**Comment by milotype** [comment]
It's easy if one knows how and if one is capable of doing it. ;-) I don't, I just translate. Please don't think of me as being ignorant because of that – I just don't know enough about programming.
I don't work with GitHub, nor with a local "branch" or whatever. I don't know to `make` nor to invoke xgettext … I only make forks of projects, download files I need, translate and send PR's.
Don't quite undertand, why the .pot file shouldn't/couldn't be automaticlly build and saved in the .po folder?

**Comment by jtanx** [comment]
> Don't quite undertand, why the .pot file shouldn't/couldn't be automaticlly build and saved in the .po folder?

Because it's a fairly large file, which would be subject to a fair amount of churn if it was automatically generated and committed, yet it's one that can be easily generated as needed. I'm already kind of annoyed with the current level of churn that gets generated each time a pofile is updated, just because of source line numbers changing ([this](https://www.berrange.com/posts/2018/11/29/improved-translation-po-file-handling-by-ditching-gettext-autotools-integration/) was an interesting read though)

This isn't really a programming issue. Like you use Poedit, xgettext is another tool that you should probably learn about, as it's a key part of how gettext functions. Or if you really don't want to use xgettext, then get Poedit to do the same thing.

Of course, I guess that all gets hidden once transifex or whatever gets used...

**Comment by milotype** [comment]
Ok, thanks. Didn't know about all that.

BTW, if **Crowdin** is still in the game ...
I like it, because it displays translations of other languages as well, enabling multilingual translators to compare translation, and make a decision when original string isn't unambiguous.

**Comment by skef** [comment]
@jtanx if it's just one largish file could we include it in the `master` branch automatic build output directory? Seems like a not-too-difficult addition and would give translators a place to grab it.

**Comment by milotype** [comment]
My current croatian .po-file is approx. 1 MB.
So I'd expect the .pot file to be even smaller.

**Comment by jtanx** [comment]
Alright, I've setup Crowdin: https://crowdin.com/project/fontforge.

I'll look into setting up the push on travis builds tomorrow. 

I'm not really satisfied with any of the GitHub integrations at the moment, so I'll hold off on doing that for now.

If this doesn't work then nbd, we can just move on and use another one of the options.

**Comment by JoesCat** [comment]
I looked at another project when we were looking at the different options, and noted basically one-liner edits, one line here, one line there....
if you're worried about churn, we're going to see plenty of it ;-P
Check the [translation commits](https://github.com/mantisbt/mantisbt/commits/master)
I think we'll see this churn regardless of whichever translator we choose.

**Comment by JoesCat** [comment]
@milotype - once again - thanks for your help - I plan to do another sweep of past and present PRs and I'm certain that I'll add you to (translation) [AUTHORS](https://github.com/fontforge/fontforge/blob/master/AUTHORS).

**Comment by milotype** [comment]
@JoesCat Great, thank you.

@jtanx Could you check in Crowdin, that the options for downloading and uploading files is enabled? These options are in the "burger"-menu icon, once one is in "FontForge.pot" file of a specific language.
As said before, I'm translating some other stuff in Crowdin. You've probably noticed there is also the possibility of setting "proofreaders". If you guys think of implementing it, I'd be happy to act as proofreader for croatian and german. (Yes, I'm planing to translate more on the german GUI …).

**Comment by jtanx** [comment]
> I looked at another project when we were looking at the different options, and noted basically one-liner edits, one line here, one line there....

That's not a fair comparison, as that's not a gettext based system. The churn I'm talking about is where the po files contain context (what line number the translation was at), which can change quite a lot without there even being any changes to the actual translations themselves.

@milotype  try now

**Comment by milotype** [comment]
@jtanx Thank you.
- downloading works fine
- you could rename the file `FontForge.pot` to `FontForge.po` (as .po file is what gets downloaded)
- AFAIK, country codes "hr_HR", "de_DE", etc. are not needed. I think its best to stick only to the main language (and only if someone thinks there is a need for "localizing", it could be added). On the other hand I don't know if Crowdin itself is handling those things.

Let me know if you think I should upload an "updated" croatian .po file as a test. My first hr-transaltion was 85%, now I'm at 95%.

**Comment by jtanx** [comment]
> AFAIK, country codes "hr_HR", "de_DE", etc. are not needed

We need them for some of the translations, e.g. to distinguish traditional vs simplified Chinese. I've updated the language mapping to just map the names that we currently use, so I think this is fixed now.

> you could rename the file FontForge.pot to FontForge.po (as .po file is what gets downloaded)

Where exactly is that? When I download the source, it has the pot extension:

![image](https://user-images.githubusercontent.com/5137410/63201166-6494a980-c0c7-11e9-8f14-e1785073a281.png)

> Let me know if you think I should upload an "updated" croatian .po file as a test. My first hr-transaltion was 85%, now I'm at 95%.

Sure, why not

**Comment by milotype** [comment]
Yes, keep countries codes where needed.

Yes it's fixed for the others. hr.po, de.po, fr.po, and en_GB (as needed).

> Where exactly is that?
Probably in the view, which "plain" translators user see. Not the "Manager" view (or whatever it may be called). Try to join the project as a translator. Then you might get a different view.

FontForge > Croatian > FontForge.pot > "Hamburger" menu > Download (see screenshot)
![translator](https://user-images.githubusercontent.com/43657314/63201798-0f38a800-c087-11e9-8d03-3a1eafa60c1c.png)

In a minute I'll send my update.

**Comment by milotype** [comment]
Also,
![FontForge pot](https://user-images.githubusercontent.com/43657314/63201932-a30a7400-c087-11e9-89a9-c2fe9d63e2de.png)

**Comment by milotype** [comment]
Ok, updated croatian. Overall it looks good I think.
Even if my hr.po file has 7257, and I downloaded a hr.po file with 7255 strings.
There seems to be some new strings, like:
- Font Editor
- org.fontforge.FontForge
and maybe some others ... others might have changed ...
(That's why I kept asking about the current .pot file.)

**Comment by JoesCat** [comment]
On August 16, 2019 02:25:53 PM Jeremy Tan wrote:

> That's not a fair comparison, as that's not a gettext based system. The
> churn I'm talking about is where the po files contain context (what
> line number the translation was at), which can change quite a lot
> without there even being any changes to the actual translations
> themselves.

Yes, the line numbers were probably okay when we did them like up to now, 
but it will generate a lot of noise as we simply edit fontforge, so 
removing the line numbers is a good move, in terms of churn.

I was thinking "churn" as in each edit generates a new replacement po file.

git type pull requests actually save new files, not the actual change like 
maybe svn which may be a small -+ type change, so a one-liner edit (in git) 
would mean another newer po file, and at 1MB each, these will add up pretty 
quickly if the edits are one line here, another, line there, another line 
here as in the example (recent) commits I saw for that example project.

**Comment by jtanx** [comment]
> git type pull requests actually save new files, not the actual change like

Please stop commenting when you don't know what you're talking about. Have a good read of this: https://stackoverflow.com/questions/8198105/how-does-git-store-files

**Comment by ctrlcctrlv** [comment]
So, Crowdin it is then? We should document that and close this. :-)

**Comment by JoesCat** [comment]
> Please stop commenting when you don't know what you're talking about. Have a good read of this: https://stackoverflow.com/questions/8198105/how-does-git-store-files

I don't see what needs corrections?
In fact, following "[a brilliant post](http://nfarina.com/post/9868516270/git-is-simpler)", I liked the comment about the 3-way merge "[Now I’ve done it. We’re going to lose all the work he did today. Sweat is beading up on my forehead.](https://nfarina.com/post/9868516270/git-is-simpler)".
Don't recall where, but I had to do that for fontforge, or libuninameslist or libspiro. Git just continued to function fine.

...anyways, getting back to your comment above...
If for example, we changed one char on a *.po, let's say, fr.po, and saved it, this fr.po file is now different from the existing git repo, so if we decide to accept the new change, the newly edited fr.po is saved into the .git directory. If we make another one char change, this newly edited fr.po is different from the git repo, so if we decide to merge this new edit, then this becomes our latest fr.po and this gets written to the .git directory. At this point, we've only just changed two characters, but we've added two new 1M (compressed) files into the .git directory.
I think @frank-trampe had to ask recently for extra space on launchpad because fontforge has a very large .git directory.
What I noticed is that our local translators have made massive edits in relatively few pull requests, while looking at 7 merges on [this page](https://github.com/mantisbt/mantisbt/commits/master), there aren't that many words edited, but plenty of PRs for such few changes.
What I was mentioning is that we're going to see the .git directory grow based on few edits often, versus massive edits done in the past.

**Comment by skef** [comment]
Just to intercede in this discussion before pies coat the walls:

```
[skef@isolator ff3]$ git checkout -b gtest
Switched to a new branch 'gtest'
[skef@isolator ff3]$ du -sk .
469244  .
[skef@isolator ff3]$ vim po/fr.po
[skef@isolator ff3]$ git diff
diff --git a/po/fr.po b/po/fr.po
index 439542a27..6808dbc8a 100644
--- a/po/fr.po
+++ b/po/fr.po
@@ -33,7 +33,7 @@ msgstr ""
 
 #: ../desktop/org.fontforge.FontForge.appdata.xml:6
 msgid "An outline font editor"
-msgstr "Un constructeur des polices"
+msgstr "Un constructeur des polices  "
 
 #: ../desktop/org.fontforge.FontForge.appdata.xml:11
 msgid ""
[skef@isolator ff3]$ git commit -a -m "test"
[gtest 4ae5c5eed] test
 1 file changed, 1 insertion(+), 1 deletion(-)
[skef@isolator ff3]$ du -sk .
469556  .
```
That's the data, on my system at least.

**Comment by JoesCat** [comment]
Thanks @skef, that is a solid answer.
Looks like approx a 312k increase for adding 2 chars, "one" commit.
git compressed the newly updated file and it's now added into the .git folder.
Seeing 7 commits on the mantisDB commit page for a handful of edits, then relatively speaking, we'd be increasing our .git archive 312k x 7 `approx 2M for a handful of edits.

> I think we'll see this churn regardless of whichever translator we choose.

was mentioned above as a comment (or as a prediction if you want to see it that way).

It was unnecessary to be writing you don't know what you're talking about, with a vague reference.
When it comes to pies on the wall, there are already quite a few.

**Comment by skef** [comment]
> It was unnecessary to be writing you don't know what you're talking about, with a vague reference.
> When it comes to pies on the wall, there are already quite a few.

@jtanx is familiar enough with git to know that you're confusing the difference between it and SVN's unit of tracking with the limit on its storage size. In fact there is no simple direct tie between the size of the file modified and the increase in the `.git` storage size:

```
[skef@isolator ff3]$ du -sk
469568  .
[skef@isolator ff3]$ git gc
Enumerating objects: 95081, done.
Counting objects: 100% (95081/95081), done.
Delta compression using up to 4 threads
Compressing objects: 100% (25529/25529), done.
Writing objects: 100% (95081/95081), done.
Total 95081 (delta 70801), reused 93258 (delta 69349)
[skef@isolator ff3]$ du -sk
419192  .
[skef@isolator ff3]$ vim po/fr.po
[skef@isolator ff3]$ git diff
diff --git a/po/fr.po b/po/fr.po
index 439542a27..31e332ebe 100644
--- a/po/fr.po
+++ b/po/fr.po
@@ -44,7 +44,7 @@ msgid ""
 msgstr ""
 "FontForge est un éditeur de polices de caractères numériques qui vous permet de créer, éditer, "
 "ou convertir une large gamme de polices incluant PostScript, TrueType, OpenType, cid-keyed, "
-"multi-master, cff, SVG et BitMap (bdf, FON, NFNT)."
+"multi-master, cff, SVG et BitMap (bdf, FON, NFNT).  "
 
 #: ../desktop/org.fontforge.FontForge.appdata.xml:16
 msgid ""
[skef@isolator ff3]$ git commit -a -m "test4"
[gtest d3d7a2e8d] test4
 1 file changed, 1 insertion(+), 1 deletion(-)
[skef@isolator ff3]$ du -sk .
419540  .
[skef@isolator ff3]$ git gc
Enumerating objects: 95087, done.
Counting objects: 100% (95087/95087), done.
Delta compression using up to 4 threads
Compressing objects: 100% (24083/24083), done.
Writing objects: 100% (95087/95087), done.
Total 95087 (delta 70804), reused 95079 (delta 70801)
[skef@isolator ff3]$ du -sk .
419444  .
[skef@isolator ff3]$ git gc --aggressive --prune=now
Enumerating objects: 95087, done.
Counting objects: 100% (95087/95087), done.
Delta compression using up to 4 threads
Compressing objects: 100% (94884/94884), done.
Writing objects: 100% (95087/95087), done.
Total 95087 (delta 71861), reused 22477 (delta 0)
[skef@isolator ff3]$ du -sk
392168  .
``` 

So your estimate was roughly accurate but the explanation you provided wasn't.

**Comment by skef** [comment]
Anyway, there is the old RCS/SCCS/SVN file ownership/locking model and the newer TeamWare/BitKeeper/Mercurial/Git fork-modify-merge model and if any battle has been won decisively over the past two decades it's that of the latter over the former.

**Comment by jtanx** [comment]
> I don't see what needs corrections?

I was kind of trying to avoid having to highlight this, but, it's mentioned pretty clearly there:

> While that's true and important on the conceptual level, it is NOT true at the storage level.

As also shown above, it's not that hard to just test this out yourself, don't forget the `git gc` step.

My patience was short, as I've lost count of the number of times you've commented based on misinformation, half-truths, or extrapolating off some unrelated idea. But in any case, this doesn't actually matter, so I'll just move on. I apologise for my terse reply before, and I'll refrain from doing that in the future.

Back on topic, yeah I guess we're going with Crowdin for now, I still need to do the auto-update/push on master build, and yeah we should document it. If this isn't working, then I'm more than happy to just switch to one of the other discussed options.

**Comment by JoesCat** [comment]
git gc, and git gc --aggressive are nice, but there is no web access to it, so it would have to be applied at the desktop.

> My patience was short, as I've lost count of the number of times you've commented based on misinformation, half-truths, or extrapolating off some unrelated idea.

It goes both ways, and more often than not, I've kept quiet about them just so we move on.

...anyways, it was FYI... and apology accepted. Time to move on.
